internal/server/v2/observation/calculation.go

```diff
@@ -17,56 +17,137 @@ package observation
 
 import (
 	"context"
+	"net/http"
 
+	pb "github.com/datacommonsorg/mixer/internal/proto"
 	pbv2 "github.com/datacommonsorg/mixer/internal/proto/v2"
 	"github.com/datacommonsorg/mixer/internal/server/cache"
+	"github.com/datacommonsorg/mixer/internal/server/resource"
 	"github.com/datacommonsorg/mixer/internal/store"
 )
 
-// CalculateObservationResponses returns a list of ObservationResponses based
-// on StatisticalCalculation formulas for the missing data in an input
-// ObservationResponse.
-func CalculateObservationResponses(
+// Computes a calculation for a variable and entity, based on a formula and input data.
+func Calculate(
 	ctx context.Context,
 	store *store.Store,
-	inputResp *pbv2.ObservationResponse,
 	cachedata *cache.Cache,
-) []*pbv2.ObservationResponse {
-	calculatedResponses := []*pbv2.ObservationResponse{}
-	for variable, variableObservation := range inputResp.ByVariable {
+	metadata *resource.Metadata,
+	httpClient *http.Client,
+	formula string,
+	variable string,
+	entity *pbv2.DcidOrExpression,
+	inputReq *pbv2.ObservationRequest,
+) (*pbv2.ObservationResponse, error) {
+	variableFormula, err := NewVariableFormula(formula)
+	if err != nil {
+		return nil, err
+	}
+	newReq := &pbv2.ObservationRequest{
+		Variable: &pbv2.DcidOrExpression{Dcids: variableFormula.StatVars},
+		Entity:   entity,
+		Date:     inputReq.GetDate(),
+		Value:    inputReq.GetValue(),
+		Filter:   inputReq.GetFilter(),
+		Select:   inputReq.GetSelect(),
+	}
+	variableResp, err := ObservationInternal(ctx, store, cachedata, metadata, httpClient, newReq)
+	if err != nil {
+		return nil, err
+	}
+	computeLeafObs(variableResp, variableFormula)
+	newResp := &pbv2.ObservationResponse{
```

Comment:
It's not clear why there are both `computeLeafObs()` and `evalExpr(variableFormula.Expr, variableFormula)`

Also this is followed by using calculatedResp to construct final response.

I feel these can be combined and simplfied?

Reply:
evalExpr is a separate function because it's recursive. I also created a separate function for computeLeafObs to only find matching obs once (and then can use these obs throughout the recursion) rather that recomputing each time a node potentially shows up. I also structured it as entity -> facet -> obs to make it easier to match facets (vs ObservationResponse, where facetId is part of the obs), but this means I have to reconstruct the response later. I moved the candidate obs out of the formula (based on your above comment) - let me know if this makes sense, or if you'd prefer I structure it differently!